server/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -72,23 +82,22 @@ private static String toLoggerName(final Class<?> cls) {
         return canonicalName != null ? canonicalName : cls.getName();
     }
 
-    public static void setThreadContext(ThreadContext threadContext) {
-        HeaderWarning.setThreadContext(threadContext);
-    }
-
-    public static void removeThreadContext(ThreadContext threadContext) {
-        HeaderWarning.removeThreadContext(threadContext);
-    }
-
     /**
-     * Logs a deprecation message, adding a formatted warning message as a response header on the thread context.
-     * The deprecation message will be throttled to deprecation log.
+     * Logs a message at the {@link #DEPRECATION} level. The message is also sent to the header warning logger,
+     * so that it can be returned to the client.
      */
-    @SuppressLoggerChecks(reason = "Safely delegates to a deprecated message")
-    public void deprecate(final String key, final String msg, final Object... params) {
-        String opaqueId = HeaderWarning.getXOpaqueId();
-        ESLogMessage deprecationMessage = new DeprecatedMessage(opaqueId, msg, params);
-        deprecationLogger.throttleLogAndAddWarning(key, deprecationMessage);
+    public DeprecationLoggerBuilder deprecate(final String key, final String msg, final Object... params) {
+        return new DeprecationLoggerBuilder().withDeprecation(key, msg, params);
     }
 
+    public class DeprecationLoggerBuilder {
```

Comment:
I wonder if we need this builder in 7.x - I removed this from my backport because there will be no compatible logger in 7.x
However I am thinking now that it might cause problems for future backports and we should keep it?
WDYT?

Reply:
Let's keep it - we've already backported more logging changes to `7.x` than we expected, and there may well be more before v8 is released.